core/services/relay/evm/chain_reader.go

```diff
@@ -23,132 +28,223 @@ type ChainReaderService interface {
+	lggr     logger.Logger
+	lp       logpoller.LogPoller
+	codec    commontypes.RemoteCodec
+	client   evmclient.Client
+	bindings Bindings
+	commonservices.StateMachine
 }
 
-// NewChainReaderService constructor for ChainReader
-func NewChainReaderService(lggr logger.Logger, lp logpoller.LogPoller, contractID common.Address, config types.ChainReaderConfig) (*chainReader, error) {
-	if err := validateChainReaderConfig(config); err != nil {
-		return nil, fmt.Errorf("%w: %w", commontypes.ErrInvalidConfig, err)
+// NewChainReaderService is a constructor for ChainReader, returns nil if there is any error
+func NewChainReaderService(lggr logger.Logger, lp logpoller.LogPoller, b Bindings, chain legacyevm.Chain, config types.ChainReaderConfig) (ChainReaderService, error) {
+	parsed := &parsedTypes{
+		encoderDefs: map[string]*codecEntry{},
+		decoderDefs: map[string]*codecEntry{},
+	}
+
+	if err := addTypes(config.ChainContractReaders, b, parsed); err != nil {
+		return nil, err
 	}
 
-	// TODO BCF-2814 implement initialisation of chain reading definitions and pass them into chainReader
-	return &chainReader{lggr.Named("ChainReader"), contractID, lp}, nil
+	c, err := parsed.toCodec()
+
+	return &chainReader{
+		lggr:     lggr.Named("ChainReader"),
+		lp:       lp,
+		codec:    c,
+		client:   chain.Client(),
+		bindings: b,
+	}, err
 }
 
 func (cr *chainReader) Name() string { return cr.lggr.Name() }
 
-func (cr *chainReader) initialize() error {
-	// Initialize chain reader, start cache polling loop, etc.
-	return nil
-}
+var _ commontypes.ContractTypeProvider = &chainReader{}
 
-func (cr *chainReader) Start(ctx context.Context) error {
-	if err := cr.initialize(); err != nil {
-		return fmt.Errorf("Failed to initialize ChainReader: %w", err)
+func (cr *chainReader) GetLatestValue(ctx context.Context, contractName, method string, params any, returnVal any) error {
+	ae, err := cr.bindings.getBinding(contractName, method, false)
+	if err != nil {
+		return err
 	}
-	return nil
-}
-
-func (cr *chainReader) Close() error { return nil }
 
-func (cr *chainReader) Ready() error { return nil }
+	if ae.evt == nil {
+		return cr.getLatestValueFromContract(ctx, contractName, method, params, returnVal)
+	}
 
-func (cr *chainReader) HealthReport() map[string]error {
-	return map[string]error{cr.Name(): nil}
+	return cr.getLatestValueFromLogPoller(ctx, contractName, method, *ae.evt, returnVal)
 }
 
-func (cr *chainReader) GetLatestValue(ctx context.Context, bc commontypes.BoundContract, method string, params any, returnVal any) error {
-	return commontypes.UnimplementedError("Unimplemented method GetLatestValue called")
+func (cr *chainReader) getLatestValueFromLogPoller(ctx context.Context, contractName, method string, hash common.Hash, returnVal any) error {
+	ae, err := cr.bindings.getBinding(contractName, method, false)
+	if err != nil {
+		return err
+	}
+
+	log, err := cr.lp.LatestLogByEventSigWithConfs(hash, ae.addr, logpoller.Finalized)
```

Comment:
We can translate ae.pending here to logpoller.Unconfirmed or logpoller.Finalized